File: sandbox/demo/idmap.py
import getpass
import logging
import shutil
import subprocess
from pathlib import Path
from typing import Iterable

from sandbox.demo.fs import write_text


def _read_subid_ranges(path: Path | str, user: str) -> list[tuple[int, int]]:
    ranges: list[tuple[int, int]] = []

    try:
        lines = Path(path).read_text().splitlines()
    except OSError as e:
        raise ValueError(f"cannot read {path}: {e}")

    for line in lines:
        # remove comments first
        parts = line.split("#", 1)[0].strip().split(":")
        if len(parts) != 3:
            continue

        if parts[0] != user:
            continue

        try:
            uid = int(parts[1])
            count = int(parts[2])
        except ValueError:
            raise ValueError(f"invalid numeric value in {path}: {line}")

        if uid < 0 or count <= 0:
            raise ValueError(f"invalid range in {path}: {line}")

        ranges.append((uid, count))

    if not ranges:
        raise ValueError(f"no subid entries for user '{user}' in {path}")

    ranges.sort(key=lambda r: r[0])

    return ranges


def _build_idmap(
    host_id: int, subid_ranges: Iterable[tuple[int, int]]
) -> list[tuple[int, int, int]]:
    idmap = [(0, host_id, 1)]
    uid = 1
    for loweruid, count in subid_ranges:
        idmap.append((uid, loweruid, count))
        uid += count
    return idmap


def _run_idmap_helper(
    helper: str,
    child_pid: int,
    idmap: list[tuple[int, int, int]],
    *,
    logger: logging.Logger,
) -> None:
    helper_path = shutil.which(helper)
    if not helper_path:
        raise RuntimeError(f"{helper} not found in PATH")

    argv = [helper_path, str(child_pid)]
    for uid, loweruid, count in idmap:
        argv += [str(uid), str(loweruid), str(count)]

    logger.debug(f"running {" ".join(argv)}")
    try:
        subprocess.run(argv, check=True, text=True, capture_output=True)
    except subprocess.CalledProcessError as e:
        msg = str(e.stderr).strip() or str(e.stdout).strip() or repr(e)
        raise RuntimeError(f"{helper} failed: {msg}") from e


def write_idmaps(
    child_pid: int, host_uid: int, host_gid: int, *, logger: logging.Logger
) -> None:
    proc = Path(f"/proc/{child_pid}")
    if not proc.exists():
        raise ValueError(f"process {child_pid} does not exist: {proc}")

    setgroups_path = proc / "setgroups"

    user = getpass.getuser()
    subuid_ranges = _read_subid_ranges("/etc/subuid", user)
    subgid_ranges = _read_subid_ranges("/etc/subgid", user)

    uidmap = _build_idmap(host_uid, subuid_ranges)
    gidmap = _build_idmap(host_gid, subgid_ranges)

    if setgroups_path.exists():
        try:
            write_text(setgroups_path, "deny\n")
            logger.debug(f"wrote {setgroups_path}: deny")
        except OSError as e:
            raise RuntimeError(f"failed to write {setgroups_path}: {e}") from e

    _run_idmap_helper("newuidmap", child_pid, uidmap, logger=logger)
    logger.debug(f"newuidmap applied: {uidmap}")

    _run_idmap_helper("newgidmap", child_pid, gidmap, logger=logger)
    logger.debug(f"newgidmap applied: {gidmap}")

File: sandbox/sandbox/idmap/idmap.go
package idmap

import (
	"fmt"
	"os"
	"os/exec"
	"os/user"
	"path/filepath"
	"strconv"
	"strings"
)

type IDRange struct {
	StartID int
	Count   int
}

type Mapping struct {
	ContainerID int
	HostID      int
	Count       int
}

func ReadSubIDRanges(path string, targetUser string) ([]IDRange, error) {
	var ranges []IDRange
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %v", path, err)
	}

	lines := strings.Split(string(data), "\n")
	for _, line := range lines {
		if idx := strings.Index(line, "#"); idx >= 0 {
			line = line[:idx]
		}
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.Split(line, ":")
		if len(parts) != 3 {
			continue
		}

		if parts[0] != targetUser {
			continue
		}

		start, err1 := strconv.Atoi(parts[1])
		count, err2 := strconv.Atoi(parts[2])
		if err1 != nil || err2 != nil || start < 0 || count <= 0 {
			return nil, fmt.Errorf("invalid numeric value or range in %s: %s", path, line)
		}

		ranges = append(ranges, IDRange{StartID: start, Count: count})
	}

	if len(ranges) == 0 {
		return nil, fmt.Errorf("no subid entries for user '%s' in %s", targetUser, path)
	}

	return ranges, nil
}

func BuildIDMap(hostID int, subidRanges []IDRange) []Mapping {
	maps := []Mapping{{ContainerID: 0, HostID: hostID, Count: 1}}
	nextContainerID := 1
	for _, r := range subidRanges {
		maps = append(maps, Mapping{
			ContainerID: nextContainerID,
			HostID:      r.StartID,
			Count:       r.Count,
		})
		nextContainerID += r.Count
	}
	return maps
}

func WriteIDMaps(childPID, hostUID, hostGID int) error {
	procPath := fmt.Sprintf("/proc/%d", childPID)

	if _, err := os.Stat(procPath); os.IsNotExist(err) && filepath.Separator == '\\' {
		return nil
	}

	currentUser, err := user.Current()
	if err != nil {
		return fmt.Errorf("failed to look up current system user: %v", err)
	}

	subUIDs, err := ReadSubIDRanges("/etc/subuid", currentUser.Username)
	if err != nil {
		return err
	}
	subGIDs, err := ReadSubIDRanges("/etc/subgid", currentUser.Username)
	if err != nil {
		return err
	}

	uidMap := BuildIDMap(hostUID, subUIDs)
	gidMap := BuildIDMap(hostGID, subGIDs)

	setGroupsPath := filepath.Join(procPath, "setgroups")
	if _, err := os.Stat(setGroupsPath); err == nil {
		if err := os.WriteFile(setGroupsPath, []byte("deny\n"), 0644); err != nil {
			return fmt.Errorf("failed to write %s: %v", setGroupsPath, err)
		}
	}

	if err := runIDMapHelper("newuidmap", childPID, uidMap); err != nil {
		return err
	}
	return runIDMapHelper("newgidmap", childPID, gidMap)
}

func runIDMapHelper(helper string, childPID int, mappings []Mapping) error {
	helperPath, err := exec.LookPath(helper)
	if err != nil {
		return fmt.Errorf("%s not found in PATH", helper)
	}

	args := []string{strconv.Itoa(childPID)}
	for _, m := range mappings {
		args = append(args, strconv.Itoa(m.ContainerID), strconv.Itoa(m.HostID), strconv.Itoa(m.Count))
	}

	cmd := exec.Command(helperPath, args...)
	if output, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("%s failed: %s (%v)", helper, strings.TrimSpace(string(output)), err)
	}
	return nil
}
